File: experimental/sso-with-skills/csharp_dotnetcore/SkillBot/Dialogs/ActivityRouterDialog.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.TraceExtensions;
using Microsoft.Bot.Schema;
using Microsoft.Extensions.Configuration;

namespace Microsoft.BotBuilderSamples.SkillBot.Dialogs
{
    /// <summary>
    /// A root dialog that can route activities sent to the skill to different sub-dialogs.
    /// </summary>
    public class ActivityRouterDialog : ComponentDialog
    {
        public ActivityRouterDialog(IConfiguration configuration)
            : base(nameof(ActivityRouterDialog))
        {
            AddDialog(new SsoSkillDialog(configuration));
            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[] { ProcessActivityAsync }));

            // The initial child Dialog to run.
            InitialDialogId = nameof(WaterfallDialog);
        }

        private async Task<DialogTurnResult> ProcessActivityAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            // A skill can send trace activities, if needed.
            await stepContext.Context.TraceActivityAsync($"{GetType().Name}.ProcessActivityAsync()", label: $"Got ActivityType: {stepContext.Context.Activity.Type}", cancellationToken: cancellationToken);

            // In this simple skill, we only handle Sso events
            if (stepContext.Context.Activity.Type == ActivityTypes.Event && stepContext.Context.Activity.Name == "Sso")
            {
                return await stepContext.BeginDialogAsync(nameof(SsoSkillDialog), cancellationToken: cancellationToken);
            }

            // We didn't get an activity type we can handle.
            await stepContext.Context.SendActivityAsync(MessageFactory.Text($"Unrecognized ActivityType: \"{stepContext.Context.Activity.Type}\".", inputHint: InputHints.IgnoringInput), cancellationToken);
            return new DialogTurnResult(DialogTurnStatus.Complete);
        }
    }
}

File: experimental/sso-with-skills/javascript_nodejs/skillBot/dialogs/activityRouterDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { ActivityTypes, MessageFactory, InputHints } = require('botbuilder');
const { ComponentDialog, WaterfallDialog, DialogTurnStatus } = require('botbuilder-dialogs');
const { SsoSkillDialog, SSO_SKILL_DIALOG } = require('./ssoSkillDialog');

const ACTIVITY_ROUTER_DIALOG = 'activityRouterDialog';
const WATERFALL_DIALOG = 'waterfallDialog';

class ActivityRouterDialog extends ComponentDialog {
    constructor() {
        super(ACTIVITY_ROUTER_DIALOG);

        this.addDialog(new SsoSkillDialog())
            .addDialog(new WaterfallDialog(WATERFALL_DIALOG, [
                this.processActivity.bind(this)
            ]));

        // The initial child Dialog to run.
        this.initialDialogId = WATERFALL_DIALOG;
    }

    async processActivity(stepContext) {
        // A skill can send trace activities, if needed.
        await stepContext.context.sendTraceActivity('ActivityRouterDialog.processActivity', `Got ActivityType: ${ stepContext.context.activity.type }`);

        // In this simple skill, we only handle SSO events.
        if (stepContext.context.activity.type === ActivityTypes.Event && stepContext.context.activity.name === 'Sso') {
            return stepContext.beginDialog(SSO_SKILL_DIALOG);
        }

        // We didn't get an activity type we can handle.
        const message = `UnrecognizedActivityType: "${ stepContext.context.activity.type }".`;
        await stepContext.context.sendActivity(MessageFactory.text(message, message, InputHints.IgnoringInput));
        return { status: DialogTurnStatus.complete };
    }
}

module.exports = {
    ActivityRouterDialog,
    ACTIVITY_ROUTER_DIALOG
};
